src/controller/tests/data_model/TestRead.cpp

```diff
@@ -295,6 +354,316 @@ void TestReadInteraction::TestReadDataVersionFilter(nlTestSuite * apSuite, void
+    //
+    // Test the application callback as well to ensure we get the right number of SubscriptionEstablishment/Termination
+    // callbacks.
+    //
+    app::InteractionModelEngine::GetInstance()->RegisterReadHandlerAppCallback(&gTestReadInteraction);
+
+    int testId = 0;
+    // Initial Read of C1A1, C1A2 and C2A1. Cache has two C1 attributes at version 0, and one C2 attribute at version 0.
+    // Expect no versions would be cached.
+    {
+        testId++;
+        ChipLogProgress(DataManagement, "\t -- Running Read with AttributeCache Test ID %d", testId);
+        app::ReadClient readClient(chip::app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(),
+                                   cache.GetBufferedCallback(), chip::app::ReadClient::InteractionType::Read);
+        readPrepareParams.mpAttributePathParamsList    = attributePathParams1;
+        readPrepareParams.mAttributePathParamsListSize = 3;
+        err                                            = readClient.SendRequest(readPrepareParams);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        ctx.DrainAndServiceIO();
+        NL_TEST_ASSERT(apSuite, delegate.mNumAttributeResponse == 3);
+        NL_TEST_ASSERT(apSuite, !delegate.mReadError);
+        Optional<DataVersion> version;
+        NL_TEST_ASSERT(apSuite, cache.GetVersion(Test::kMockEndpoint2, Test::MockClusterId(3), version) == CHIP_NO_ERROR);
+        NL_TEST_ASSERT(apSuite, !version.HasValue());
+        delegate.mNumAttributeResponse = 0;
+    }
+    // Second read of C1A* and C2A1. We cannot use the stored data versions in the cache since our subsequent read is 'broader' then
+    // the previous read that resulted in the data in the cache. Expect cache C1 version
+    {
+        testId++;
+        ChipLogProgress(DataManagement, "\t -- Running Read with AttributeCache Test ID %d", testId);
+        app::ReadClient readClient(chip::app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(),
+                                   cache.GetBufferedCallback(), chip::app::ReadClient::InteractionType::Read);
+        readPrepareParams.mpAttributePathParamsList    = attributePathParams2;
+        readPrepareParams.mAttributePathParamsListSize = 2;
+        err                                            = readClient.SendRequest(readPrepareParams);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        ctx.DrainAndServiceIO();
+        NL_TEST_ASSERT(apSuite, delegate.mNumAttributeResponse == 6);
+        NL_TEST_ASSERT(apSuite, !delegate.mReadError);
+        Optional<DataVersion> version;
+        NL_TEST_ASSERT(apSuite, cache.GetVersion(Test::kMockEndpoint2, Test::MockClusterId(3), version) == CHIP_NO_ERROR);
+        NL_TEST_ASSERT(apSuite, version.HasValue() && (version.Value() == 0));
+        delegate.mNumAttributeResponse = 0;
+    }
+
+    // Third read of C1A1 and C2A1. It would use the stored data versions in the cache since our subsequent read's C1A1 path
```

Comment:
You're reading C1A1 C2A1 **AND** C2A2 right?

Reply:
right, updated. thanks